### PR TITLE
feat: Implement tests

### PR DESCRIPTION
**Pull request by martinalbert**

### CODE REVIEW

File: butter_cms/content_field.py [hunk 1]
```diff
@@ -12,3 +12,13 @@ def get(self, keys=None, params=None):
             params = {}
         params['keys'] = ','.join(keys)
         return self.api_get(params=params)
+    
+    def get(self, slug, fields=None, params=None):
```
Comment (Toreno96) [inline butter_cms/content_field.py:16]: Doesn't it override the previous `def get`?

File: butter_cms/content_field.py [hunk 2]
```diff
@@ -12,3 +12,13 @@ def get(self, keys=None, params=None):
...
+        if not params:
+            params = {}
+        if fields:
+            for field in fields:
+                key = "fields.{}".format(field)
```
Comment (Toreno96) [inline butter_cms/content_field.py:21]: Can we use f-string here?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+name: Python Tests
+
+on: 
+  workflow_call:
+
+jobs:
+  test:
+    runs-on: ubuntu-24.04
```
Comment (Toreno96) [inline .github/workflows/test.yml:8]: This should be `ubuntu-22.04`, right?

**Review by Toreno96 — CHANGES_REQUESTED** [review]

File: butter_cms/content_field.py [hunk 4]
```diff
@@ -12,3 +12,13 @@ def get(self, keys=None, params=None):
             params = {}
         params['keys'] = ','.join(keys)
         return self.api_get(params=params)
+    
+    def getCollection(self, slug, fields=None, params=None):
```
Comment (Toreno96) [inline butter_cms/content_field.py:16]: Could we rename this using snake_case, as a common convention in Python suggests?

Comment (martinalbert) [reply butter_cms/content_field.py:16]: of course, my bad 🤦🏻‍♂️

Comment (martinalbert) [reply butter_cms/content_field.py]: done ✅

File: README.md [hunk 5]
```diff
@@ -1,7 +1,15 @@
...
 
+[![PyPI version](https://badge.fury.io/py/buttercms-python.svg)](https://badge.fury.io/py/buttercms-python)
+[![TestPyPI version](https://img.shields.io/pypi/v/buttercms-python?label=TestPyPI)](https://test.pypi.org/project/buttercms-python/)
+[![Downloads](https://pepy.tech/badge/buttercms-python)](https://pepy.tech/project/buttercms-python)
+[![Downloads](https://pepy.tech/badge/buttercms-python/month)](https://pepy.tech/project/buttercms-python)
+[![Downloads](https://pepy.tech/badge/buttercms-python/week)](https://pepy.tech/project/buttercms-python)
+
 Python Library for ButterCMS API. 
 
+We officially support Python versions 3.8 to 3.12. While compatibility with other Python 3.x versions is possible, as we currently do not utilize any version-specific features, it is not guaranteed.
```
Comment (Toreno96) [inline README.md]: We could put this in the `NOTE` (or `IMPORTANT`?) alert box: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

WDYT?

File: requirements.txt [hunk 6]
```diff
@@ -1,8 +1,11 @@
...
+certifi==2024.2.2; python_version >= '3.8'
+charset-normalizer==3.3.2; python_version >= '3.8'
+idna==3.7; python_version >= '3.8'
+packaging==24.0; python_version >= '3.8'
+pyproject_hooks==1.1.0; python_version >= '3.8'
+requests==2.31.0; python_version >= '3.8'
+urllib3==2.2.1; python_version >= '3.8'
+importlib_metadata==7.1.0; python_version == '3.8' or python_version == '3.9'
+tomli==2.0.1; python_version == '3.8' or python_version == '3.9' or python_version == '3.10'
+zipp==3.18.1; python_version == '3.8' or python_version == '3.9'
```
Comment (Toreno96) [inline requirements.txt:19]: Are those not needed for Python > 3.9 (> 3.10 in the case of `tomli`)?

File: butter_cms/content_field.py [hunk 7]
```diff
@@ -12,3 +12,13 @@ def get(self, keys=None, params=None):
...
+    
+    def get_collection(self, slug, fields=None, params=None):
+        if not params:
+            params = {}
+        if fields:
+            for field in fields:
+                key = f"fields.{field}"
+                params[key] = fields[field]
+
+        return self.api_get(slug, params=params)
```
Comment (Toreno96) [inline butter_cms/content_field.py]: Why was this added? I don't see it being used anywhere:
```
$ rg get_collection
butter_cms/content_field.py
16:    def get_collection(self, slug, fields=None, params=None):
```

File: requirements.txt (hunk 6 above)
Comment (martinalbert) [reply requirements.txt:19]: yes, python 3.10 and higher requires only first 8

File: butter_cms/content_field.py (hunk 7 above)
Comment (martinalbert) [reply butter_cms/content_field.py]: The method `get` is checking for all collections with ability to filter based on `keys` (this is used without `slug` - you can see uses of this in [docs](https://buttercms.com/docs/api) if you search for `content/?keys=` on the page), but there is no support for checking specific collection (docs: https://buttercms.com/docs/api/#filtering-a-collection) that is able to filter items in specific collection.

Comment (Toreno96) [reply butter_cms/content_field.py]: Makes sense. I think we should add unit tests for that, tho. I think we also should consider adding docstrings (those could be added for everything in the SDK tbh, so maybe we should consider this as a separate task and PR cc @ViolanteCodes).

**Review by Toreno96 — CHANGES_REQUESTED** [review]
I was able to finally review the unit tests themselves, and I have bad news: I think all of those require reimplementation.

Let's see an example, `TestListAuthor` 👇

We mock `Author.all` (the function on the highest possible level of abstraction) to return `list_authors()`:

https://github.com/ButterCMS/buttercms-python/blob/e9741753f810cebca47b8196af29de94a1785f87/butter_cms/tests/test_author.py#L8-L11

So we don't really test any internals of the `Author` class other than the initialization of the class, but even that is not much because the only thing the initialization does is assigning/defining a few member variables, which we then don't use at all because of the mock.

So:

```python
    @patch('butter_cms.author.Author.all')
    def test_all(self, mock_all):
        mock_all.return_value = list_authors()
        author = Author('<demo-token>')
        response = author.all()
        …
```

This is effectively the same as:

```python
    def test_all(self):
        response = list_authors()
        …
```

Furthermore, the whole test is effectively the same as:

```python
    def test_all(self):
        response = {
            'data': [
                {
                    'first_name': 'John', 
                    'last_name': 'Doe', 
                    'slug': 'john-doe'
                },
                {
                    'first_name': 'Jane', 
                    'last_name': 'Doe', 
                    'slug': 'jane-doe'
                }
            ]
        }

        self.assertIsNotNone(response)
        self.assertIn('data', response)

        data = response['data']
        self.assertIsNotNone(data)
        self.assertIsInstance(data, list)
        self.assertGreater(len(data), 0)

        author_data = data[0]
        self.assertEqual('John', author_data.get('first_name'))
        self.assertIsNone(author_data.get('recent_posts'))
```

I think at this point you can see, that this is kind of the same as `assert true == true`; which is not a good unit test 😅

As such, I think all the current tests have to be removed and rewritten.

I was thinking what could be a good approach. We can't mock `all()`  or `get()` or similar because of the above reasons.
I don't think we can mock `Client.api_get()` either because in a significant number of cases `all()` or `get()` is just passing through the result of `api_get`, and only in some cases it does a _bit_ of additional work, so mocking `api_get()` would be the same as mocking `all()` or `get()` for the most part.

I believe a good approach could be to rely on https://pypi.org/project/responses/ to mock responses from `request` library (which we use inside the `api_get`) without mocking any method of `Client` and derivatives themselves. Something like (pseudocode):

```python
    def test_all(self):
        # mock an appropriate response(s) using `responses`
        …
        # then
        author = Author('<demo-token>')
        response = author.all()
        …
```

This is not perfect either because unit tests aren't supposed to know that the tested function is using library X or Y internally, but I don't see any better solution, unfortunately.

@martinalbert WDYT?

**Comment by martinalbert** [comment]
@Toreno96 thank you for the review, I understand your concerns and I don't like this approach as well.

I tried the `responses` package as well as `vcr` and I'm thinking if it wouldn't make more sense to use the VCR while we're at it. I tested it on one test and you can see the results in this commit: 62f5980a6dc5423e95f1e76b75e9f0a9a40b3c28

**Comment by martinalbert** [comment]
@Toreno96 Just to let you know why I tried also the `VCR` along the `responses`
- They both solve the mentioned concern of superficial unit tests - intercepting underlying HTTP traffic instead of method calls,
- It feels like similar amount of work would be needed to make the tests better,
- VCR removes the need to create mocks manually and uses more realistic data

File: butter_cms/tests/test_author.py [hunk 8]
```diff
@@ -47,44 +47,44 @@ def test_all_include_recent_posts(self):
         self.assertGreater(len(data), 0)
 
         author_data = data[0]
-        self.assertEqual('John', author_data.get('first_name'))
+        self.assertEqual('SDK', author_data.get('first_name'))
```
Comment (Toreno96) [inline butter_cms/tests/test_author.py]: Could we compare full body, not just single field? For every test for every type of data (authors, category, content_field, etc).

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (Toreno96) [inline butter_cms/tests/fixtures/authors/test_author_all.json:1]: Does VCR know this is the result for `DEMO_TOKEN` environment variable, or would it work for any token, including `invalid_token`?

File: butter_cms/content_field.py [hunk 9]
```diff
@@ -7,8 +7,21 @@ def __init__(self, auth_token):
         Client.__init__(self, auth_token)
         self.path = 'content/'
 
-    def get(self, keys=None, params=None):
+    def get(self, keys, params=None):
```
Comment (Toreno96) [inline butter_cms/content_field.py:10]: Why `keys` is now required?

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (martinalbert) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: This recording points to environment tied with DEMO_TOKEN as it was recorded with that token, it will only work with token pointing to identical environment.

File: butter_cms/content_field.py (hunk 9 above)
Comment (martinalbert) [reply butter_cms/content_field.py:10]: Because as you can see in tests, when no keys are provided, result is empty. This will at least guide developer that this should be used with some keys.

File: butter_cms/tests/test_author.py (hunk 8 above)
Comment (martinalbert) [reply butter_cms/tests/test_author.py]: Added this in 929c6feb46613cc6663ccf718d7af6c3f74fda52

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (Toreno96) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: In that case, does it require any special setup in the CI/CD?

File: butter_cms/content_field.py (hunk 9 above)
Comment (Toreno96) [reply butter_cms/content_field.py:10]: > This will at least guide developer that this should be used with some keys.

Hmm, maybe we could indicate it better by using type annotations, WDYT?

**Review by Toreno96 — CHANGES_REQUESTED** [review]
Looks much better now!

Please see:
https://github.com/ButterCMS/buttercms-python/pull/11#discussion_r1628116765
https://github.com/ButterCMS/buttercms-python/pull/11#discussion_r1628118585

Also, I see that in a few places you're defining variables like `f = data['navigation_menu_item']`, `d1 = data[0]`, etc.

__Could you rename them all to something more meaningful?__
Single-letter variables are unclear and difficult to search, and it's not like we have a character limit or anything.

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (martinalbert) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: No, VCR handles this by matching this recording to the test case.

File: butter_cms/content_field.py (hunk 9 above)
Comment (martinalbert) [reply butter_cms/content_field.py:10]: This is similar as indicating that `search` method on `author` or `post` requires `query`. It would be definitely better to use type annotations but IMO that would be outside the scope of this PR.

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (Toreno96) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: I mean, there's `os.getenv('DEMO_TOKEN')` in the code, will VCR set up this environment variable?

File: butter_cms/content_field.py (hunk 9 above)
Comment (Toreno96) [reply butter_cms/content_field.py:10]: Ok, makes sense.

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (martinalbert) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: It will yes, but even if it's going to be empty in CI/CD, VCR will ignore it and mock or intercept the http traffic with this recording based on matching parameters like method, host, path and query. VCR also offers filtering of query parameters that would be problematic like in this case the token by using `filter_query_parameters`. You can see this in https://github.com/ButterCMS/buttercms-python/blob/feat/implement-tests/butter_cms/tests/test_vcr_config.py

Comment (martinalbert) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: I also made sure that tests are not using any outgoing http calls in this commit: f2362837ce9e2de4451877b8617f184651209e82

Comment (Toreno96) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: > I also made sure that tests are not using any outgoing http calls in this commit: [f236283](https://github.com/ButterCMS/buttercms-python/commit/f2362837ce9e2de4451877b8617f184651209e82)

Wow, that's clever! I didn't even know it's possible to block outgoing traffic on the Python side, thanks!



> It will yes, but even if it's going to be empty in CI/CD, VCR will ignore it and mock or intercept the http traffic with this recording based on matching parameters like method, host, path and query.

I'll be honest, I'm confused. Originally we've discussed:

>> Does VCR know this is the result for `DEMO_TOKEN` environment variable, or would it work for any token, including `invalid_token`?
>
> This recording points to environment tied with DEMO_TOKEN as it was recorded with that token, it will only work with token pointing to identical environment.

I find your most recent comment contradicting.

If both are true:
- the recording points to an environment tied with `DEMO_TOKEN`
- if `DEMO_TOKEN` is empty in CI/CD, VCR will ignore it

Then doesn't it mean the recording is not actually tied to the `DEMO_TOKEN`? I mean, if `DEMO_TOKEN` is missing, the tests will behave exactly the same as if `DEMO_TOKEN` is defined, do I understand correctly?

In that case, we could make `DEMO_TOKEN==""`, `DEMO_TOKEN=="abc"`, `DEMO_TOKEN=="invalid token"`, and all of them will work for the same (return a valid response for the test that uses the recording returning a valid response). Correct? Or am I misunderstanding?

Comment (martinalbert) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: I understand the confusion, VCR have different modes for recording the http traffic. Currently it records the traffic only once and for next calls, it uses mechanism to match the recordings to the test cases.

One part of that mechanism is recording name defined at the top of test case like [here](https://github.com/ButterCMS/buttercms-python/blob/feat/implement-tests/butter_cms/tests/test_author.py#L13)

After it takes this recording, it will validate that recording really matches the http call that we're doing. This is done by checking the method, host, path and query.

Since we filter the api_key from query (to not include sensitive information) - VCR does not differentiate between valid and invalid token. It uses path to recording and other matching parameters to validate the recording.

> I mean, if DEMO_TOKEN is missing, the tests will behave exactly the same as if DEMO_TOKEN is defined, do I understand correctly?

Yes. The difference will be spotted only if you will re-record the traffic.

> This recording points to environment tied with DEMO_TOKEN as it was recorded with that token, it will only work with token pointing to identical environment.

By this I meant that recording that you pointed out was created with valid token pointing to one environment, so if you choose to re-record the traffic (for example with `record_mode='all'`) it will have to be pointing to identical environment to let the tests pass.

Comment (Toreno96) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: @martinalbert Thank you for the detailed explanation, all is clear now! Sorry for the delayed response, but I have a lot on my plate recently.

It's not ideal that the unit tests work around the authentication, but I also understand there's no obvious solution to that.

I'm thinking if we could at least document it somewhere (by copy-pasting your above comment and rephrasing it a bit). Doing it above every `self.token = os.getenv('DEMO_TOKEN')` in each file wouldn't be practical, so I'm open to suggestions. Maybe we could extract each `self.token = os.getenv('DEMO_TOKEN')` into a common function and put the comment inside it? Let me know your thoughts.

Apart from that, all LGTM, so this is one last thing to do :)

File: butter_cms/tests/test_vcr_config.py [hunk 10]
```diff
@@ -1,3 +1,8 @@
+import socket
+def guard(*args, **kwargs):
+    raise Exception("Blocked network access. Use VCR cassettes instead.")
+socket.socket = guard
+
```
Comment (Toreno96) [inline butter_cms/tests/test_vcr_config.py]: Out of curiosity, would it also block connection to a database/redis/etc if we use one, considering that we're blocking the whole socket?

I know we don't use them here, but I'm curious if similar solution can be used in e.g. a sophisticated Django app or not.

Comment (martinalbert) [reply butter_cms/tests/test_vcr_config.py]: @Toreno96 This file will be only executed during tests, I tested it with django starter and moved tests into root outside of main module for possible mistakes in build commands. 👍🏻

Comment (Toreno96) [reply butter_cms/tests/test_vcr_config.py]: What I meant is would it block the connection to a database/redis/etc __in tests__. Sorry for the confusion.

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (Toreno96) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: Hmm, I don't see a mention of authentication in the [newly introduced explanation](https://github.com/ButterCMS/buttercms-python/pull/11/commits/e98745abac072ce63095d6729401f34a5c386d58#diff-44fdcc1c98a7701c52d948e91dcc892d6b0ee2e1847a8d649ffe8fbd4e96d897R49-R66), which was my main focus in the suggestion.

File: butter_cms/tests/test_vcr_config.py (hunk 10 above)
Comment (martinalbert) [reply butter_cms/tests/test_vcr_config.py]: Yes, it would block all network connections.

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (martinalbert) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: @Toreno96 I tried to rephrase it in 0644e26e60516e4bed3fe46c85873f4a37d7b055, let me know if it's better.

File: butter_cms/tests/test_vcr_config.py (hunk 10 above)
Comment (Toreno96) [reply butter_cms/tests/test_vcr_config.py]: Ok, thanks.

File: butter_cms/tests/fixtures/authors/test_author_all.json
Comment (Toreno96) [reply butter_cms/tests/fixtures/authors/test_author_all.json:1]: Much better, thank you! Great job!

**Review by Toreno96 — APPROVED** [review]
LGTM

**Review by ViolanteCodes — APPROVED** [review]
